src/facet.js

```diff
@@ -36,26 +36,33 @@ class Facet extends Mark {
     for (const facetKey of facetsKeys) {
       marksIndexByFacet.set(facetKey, new Array(this.marks.length));
     }
+    let facetsExclude;
     for (let i = 0; i < this.marks.length; ++i) {
       const mark = this.marks[i];
-      const markFacets = mark.data === this.data ? facetsIndex : undefined;
-      const {index, channels} = mark.initialize(markFacets);
+      const {facet} = mark;
+      const markFacets = facet === "auto" ? mark.data === this.data ? facetsIndex : undefined
+        : facet === "include" ? facetsIndex
+        : facet === "exclude" ? facetsExclude || (facetsExclude = facetsIndex.map(f => Uint32Array.from(difference(index, f))))
+        : undefined;
+      if (markFacets && range(mark.data).length !== index.length) {
+        throw new Error("faceted mark data must match facet data length");
+      }
```

Comment:
`range(mark.data).length` is inefficient test in the common case where mark.data is already an array. Also, this test isn’t strictly related to the _facet_ option, and might be overly strict. Maybe we shouldn’t add this?

```suggestion
```

Reply:
My idea here was to raise a flag before users tried to pass data they had reduced themselves and wondered why it didn't magically follow facets, but the test isn't fool-proof.
Shouldn't we have an all-purpose function for length? Or is it simply `arrayify(data).length`?